solo/methods/sdbyol.py

```diff
@@ -0,0 +1,337 @@
+
+        return parent_parser
+
+    @property
+    def learnable_params(self) -> List[dict]:
+        """Adds projector and predictor parameters to the parent's learnable parameters.
+
+        Returns:
+            List[dict]: list of learnable parameters.
+        """
+
+        extra_learnable_params = [
+            {"params": self.projector.parameters()},
+            {"params": self.predictor.parameters()},
+            {"params": self.embedder.parameters()},
+            {"name": 'classifier_y',
+              "params": self.linear_y.parameters(),
+              "lr": self.classifier_lr,
+              "weight_decay": 0
+              },
+        ]
+        return super().learnable_params + extra_learnable_params
+
+    @property
+    def momentum_pairs(self) -> List[Tuple[Any, Any]]:
+        """Adds (projector, momentum_projector) to the parent's momentum pairs.
+
+        Returns:
+            List[Tuple[Any, Any]]: list of momentum pairs.
+        """
+
+        extra_momentum_pairs = [(self.projector, self.momentum_projector), (self.embedder, self.momentum_embedder)]
+        return super().momentum_pairs + extra_momentum_pairs
+
+    def forward(self, X: torch.Tensor, *args, **kwargs) -> Dict[str, Any]:
+        """Performs forward pass of the online backbone, projector and predictor.
+
+        Args:
+            X (torch.Tensor): batch of images in tensor format.
+
+        Returns:
+            Dict[str, Any]: a dict containing the outputs of the parent and the projected features.
+        """
+
+        out = super().forward(X, *args, **kwargs)
+        emb = self.embedder(out['feats'])
+        y = self.softmax(emb)
+        z = self.projector(y)
+        p = self.predictor(z)
+        return {**out, "z": z, "p": p, "logits": logits, "y": y}
+
+    def _class_step(self, X, targets, classifier):
+        logits = classifier(X)
+
+        loss = F.cross_entropy(logits, targets, ignore_index=-1)
+        acc1, acc5 = accuracy_at_k(logits, targets, top_k=(1, 5))
+        return {"loss": loss, "acc1": acc1, "acc5": acc5}
+
+    def _shared_step(
+        self, feats: List[torch.Tensor], momentum_feats: List[torch.Tensor], targets: List[torch.tensor]
+    ) -> torch.Tensor:
+
+        embs = [self.embedder(f) for f in feats]
+        Y = [self.softmax(e) for e in embs]
+        Z = [self.projector(y) for y in Y]
+        P = [self.predictor(z) for z in Z]
+
+        # forward momentum backbone
+        with torch.no_grad():
+            embs_momentum = [self.momentum_embedder(f) for f in momentum_feats]
+            Y_momentum = [self.momentum_softmax(e) for e in embs_momentum]
+            Z_momentum = [self.momentum_projector(y) for y in Y_momentum]
+
+        # ------- negative consine similarity loss -------
+        neg_cos_sim = 0
+        for v1 in range(self.num_large_crops):
+            for v2 in np.delete(range(self.num_crops), v1):
+                neg_cos_sim += byol_loss_func(P[v2], Z_momentum[v1])
```

Comment:
does `+=` work now with pytorch and autodiff? I thought we had to write always `neg_cos_sim = neg_cos_sim + ...`

Reply:
is `self.num_crops` the same as `self.num_large_crops` here?
so the list is for different views??

Reply:
It does work as far as I can tell, it does train. However, maybe I am missing something.

Yes, num_crops is generally the same as num_large_crops. But, I guess that they could be different.

This for loop is a trick that they implemented for handling multi-cropping. Maybe not the sexiest implementation :P